Switch from PhantomJS to Headless Chrome

rowanmanning opened the pull request:
This PR replaces PhantomJS with Headless Chrome. This will hopefully go into the 5.x branch and be released later with the rest of that stuff. I've opened up several issues in the [5.0 milestone](https://github.com/pa11y/pa11y/milestone/7) to address the `TODO` items that appear in this PR.

If you want to do a "first-pass" review, I'd recommend looking at the README and MIGRATION file changes mostly.

Sorry for the huge amount of code, I think after this I may release a beta so that people can try it out. I think it'll be easier to just [look at the code on the branch](https://github.com/pa11y/pa11y/tree/headless-chrome) rather than review the changes as most of the JS has had to be completely rewritten.

## Questions we've been asked (or we anticipate)

### Why Node.js 8+ only?
We wanted to switch to using the most modern development methods we can. We also wanted to use [Puppeteer](https://github.com/GoogleChrome/puppeteer), which requires Node.js 7.10+. We'll be keeping Pa11y 4.x maintained for quite some time (possibly as far as April 2019 when Node.js 6.x maintenance ends) and will update everyone when we decide on an end-of-support date.

---

(I've just noticed this is #304 😁 definitely not a status code that describes this work)

andrewmee commented:
All the documentation looks good 👍  

Will try to get some time to dive in to the code for review soon (if someone doesn't get to that first)

hollsk commented:
I've set myself a reminder to have a look tomorrow, but the more the merrier 😁

rowanmanning commented on lib/action.js:
```diff
@@ -1,72 +1,57 @@
-'use strict';
```
Oh well spotted, I would have forgotten! Thanks for the review, Lily :)

TheJaredWilcurt replied:
Why not just have everything pass a linter before allowing it to be merged?

* https://eslint.org/docs/rules/strict

rowanmanning replied:
@TheJaredWilcurt: we do use a linter across the projects, but the contribution guidelines outline some basic principals that contributors can stick to in order to make their lives easier later.

@Lily2point0: I've decided I'm just gonna add the `'use strict';` back in 😁

hollsk commented on MIGRATION.md:
```diff
@@ -8,17 +8,67 @@ Pa11y's API changes between major versions. This is a guide to help you make the
+
+### Node.js Support
+
+Pa11y 5.0 now only supports Node.js v8.0.0 and higher, you'll need to upgrade to be able to use the latest versions of Pa11y.
```
I'd change "now only supports" to "only supports" to remove any kind of timeline ambiguity (like, "when is now?" philosoraptor.jpg)

hollsk commented on MIGRATION.md:
```diff
@@ -8,17 +8,67 @@ Pa11y's API changes between major versions. This is a guide to help you make the
+The command-line interface in 5.0 is similar to 4.0, but there are a few key changes.
+
+  - The `--verify-page` flag has been removed, as page verification can be achieved with screen-shots or observing the debug output
+  - The `--htmlcs` flag has been removed, you can no longer configure the version of HTML CodeSniffer that Pa11y uses (TODO reconsider this)
```
In two minds. On the one hand it'd be useful for people who want to upgrade/downgrade htmlcs (could this have saved us during our second-to-last release? 🤔 ). 

On the other hand maybe it'd be better to come up with an API of sorts so people can plug in $ENGINE and have it produce standard output.

rowanmanning replied:
Yeah I think _maybe_ it's easier to remove this and readdress if somebody needs to switch out Phantom. Technically they still can but just with their own fork/build of Pa11y (if the need is dire and we don't have engine switching figured out.

hollsk commented on MIGRATION.md:
```diff
@@ -8,17 +8,67 @@ Pa11y's API changes between major versions. This is a guide to help you make the
+Configuration options have had an update between 4.0 and 5.0:
+
+  - The `allowedStandards` option has been removed. This can still be set by manually adding entries to `pa11y.allowedStandards`
+  - The `beforeScript` option has been removed in favour of using actions
```
Maybe add a link to #228 with a note saying 'if something's broke, tell us there'? 😁

hollsk commented on MIGRATION.md:
```diff
@@ -8,17 +8,67 @@ Pa11y's API changes between major versions. This is a guide to help you make the
+
+### Miscellaneous Changes
+
+  - A full scheme is now required in Pa11y URLs, you can no longer run `pa11y example.com`, you need to use `pa11y http://example.com/` (TODO reconsider this)
```
I don't have a problem with this change. I always use the scheme anyway, IIRC - maybe other reviewers have an opinion?

rowanmanning replied:
Yeah I just feel like it might be an unnecessary removal, and cause loads of questions. It's only a couple of lines of code to add it back in and maintain backwards-compatibility with more of the CLI.

hollsk commented on TROUBLESHOOTING.md:
```diff
@@ -1,82 +1,64 @@
-If you do need to test the contents of an iframe, run Pa11y against the iframe source URL directly.
 
-### PhantomJS fails to run and returns exit code 127
+TODO verify that this is still an issue with Headless Chrome
```
We won't know if headless chrome has similar problems until it's being used out in the wild - even if it does have similar problems, they won't really look like this one. 

Maybe we should leave this section alone so that 4.x users can still refer to it, until 4.x is retired

hollsk commented on TROUBLESHOOTING.md:
````diff
@@ -86,107 +68,37 @@ pa11y({
-    }
-});
-```
+We recommend using [actions][actions].
````
This would be a good place to plug #228 again 😁

rowanmanning commented:
OK @hollsk: I've updated the docs based on all your changes/comments.

@Lily2point0: I've added `use strict` back in, I'm not 100% sure why I removed it anyway!

Thanks so much all of you for the reviews, your feedback is really helpful :)
